File: frontend/src/api/conversations.ts
import type { ConversationDetail, ConversationSummary } from '../types'
import { apiFetch } from './client'

export function listConversations() {
  return apiFetch<ConversationSummary[]>('/api/conversations')
}

export function createConversation(title = 'New chat') {
  return apiFetch<ConversationSummary>('/api/conversations', {
    method: 'POST',
    body: JSON.stringify({ title }),
  })
}

export function getConversation(id: string) {
  return apiFetch<ConversationDetail>(`/api/conversations/${id}`)
}

export function deleteConversation(id: string) {
  return apiFetch<void>(`/api/conversations/${id}`, { method: 'DELETE' })
}

export function updateConversationTitle(id: string, title: string) {
  return apiFetch<ConversationSummary>(`/api/conversations/${id}`, {
    method: 'PATCH',
    body: JSON.stringify({ title }),
  })
}

File: backend/app/routers/conversations.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas import (
    ConversationCreate,
    ConversationDetail,
    ConversationSummary,
    ConversationUpdate,
)
from app.services import conversation_service

router = APIRouter(prefix="/conversations", tags=["conversations"])


@router.get("", response_model=list[ConversationSummary])
def list_conversations(db: Session = Depends(get_db)):
    return conversation_service.list_conversations(db)


@router.post("", response_model=ConversationSummary, status_code=201)
def create_conversation(
    data: ConversationCreate | None = None,
    db: Session = Depends(get_db),
):
    return conversation_service.create_conversation(db, data)


@router.get("/{conversation_id}", response_model=ConversationDetail)
def get_conversation(conversation_id: str, db: Session = Depends(get_db)):
    conversation = conversation_service.get_conversation(db, conversation_id)
    if not conversation:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return conversation


@router.patch("/{conversation_id}", response_model=ConversationSummary)
def update_conversation(
    conversation_id: str,
    data: ConversationUpdate,
    db: Session = Depends(get_db),
):
    conversation = conversation_service.get_conversation(db, conversation_id)
    if not conversation:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return conversation_service.update_conversation(db, conversation, data)


@router.delete("/{conversation_id}", status_code=204)
def delete_conversation(conversation_id: str, db: Session = Depends(get_db)):
    conversation = conversation_service.get_conversation(db, conversation_id)
    if not conversation:
        raise HTTPException(status_code=404, detail="Conversation not found")
    conversation_service.delete_conversation(db, conversation)
